benches/nexmark/generator/person_generator.rs

```diff
@@ -0,0 +1,246 @@
+// Copyright 2022 the DBSP contributors.
+// SPDX-License-Identifier: MIT
+
+//! Generates people for the Nexmark streaming data.
+//!
+//! API based on the equivalent [Nexmark Flink PersonGenerator API](https://github.com/nexmark/nexmark/blob/v0.2.0/nexmark-flink/src/main/java/com/github/nexmark/flink/generator/model/PersonGenerator.java).
+
+use rand::{seq::SliceRandom, Rng};
+
+use super::strings_generator::next_string;
+use crate::model::{DateTime, Id, Person};
+
+// Keep the number of states small so that the example queries will find
+// results even with a small batch of events.
+const US_STATES: &[&str] = &["AZ", "CA", "ID", "OR", "WA", "WY"];
+
+const US_CITIES: &[&str] = &[
+    "Phoenix",
+    "Los Angeles",
+    "San Francisco",
+    "Boise",
+    "Portland",
+    "Bend",
+    "Redmond",
+    "Seattle",
+    "Kent",
+    "Cheyenne",
+];
+
+const FIRST_NAMES: &[&str] = &[
+    "Peter", "Paul", "Luke", "John", "Saul", "Vicky", "Kate", "Julie", "Sarah", "Deiter", "Walter",
+];
+
+const LAST_NAMES: &[&str] = &[
+    "Shultz", "Abrams", "Spencer", "White", "Bartels", "Walton", "Smith", "Jones", "Noris",
+];
+
+// TODO(absoludity): add GeneratorConfig rather than hard-coding.
+const PERSON_PROPORTION: usize = 3;
+const TOTAL_PROPORTION: usize = 10;
+const NUM_ACTIVE_PEOPLE: usize = 2;
+const FIRST_PERSON_ID: usize = 0;
+
+// Generate and return a random person with next available id.
+// TODO(absoludity): Update to take GeneratorConfig.
+pub fn next_person<R: Rng + ?Sized>(next_event_id: Id, rng: &mut R, timestamp: u64) -> Person {
+    // TODO(absoludity): Figure out the purpose of the extra field - appears to be
+    // aiming to adjust the number of bytes for the record to be an average, which will
+    // need slightly different handling in Rust.
+    // int currentSize =
+    //     8 + name.length() + email.length() + creditCard.length() + city.length() + state.length();
+    // String extra = nextExtra(random, currentSize, config.getAvgPersonByteSize());
+
+    // Convert timestamp milliseconds into second and nano-seconds
+    let secs = timestamp / 1000;
+    let nsecs = (timestamp % 1000) * 1000;
+    Person {
+        id: last_base0_person_id(next_event_id) + FIRST_PERSON_ID,
+        name: next_person_name(rng),
+        email_address: next_email(rng),
+        credit_card: next_credit_card(rng),
+        city: next_us_city(rng),
+        state: next_us_state(rng),
+        date_time: DateTime::from_timestamp(secs.try_into().unwrap(), nsecs.try_into().unwrap()),
+        extra: String::new(),
+    }
+}
+
+/// Return a random person id (base 0).
```

Comment:
Base zero?

Reply:
I've just copied the doc-strings from the original Java implementation. In this case, it's from:

https://github.com/nexmark/nexmark/blob/54974ef36a0d01ef8ebc0b4ba39cfc50136af0f6/nexmark-flink/src/main/java/com/github/nexmark/flink/generator/model/PersonGenerator.java#L69

I'm not (yet) sure what the original author meant, but *assume* it's related to the `epoch * PERSON_PROPORTION + offset` (see `last_base0_person_id` below). I'll need to look more into the original implementation background (don't think it's mentioned in the paper).